File: javascript/grade-school/grade-school.js
export class GradeSchool {
  constructor() {
    this._db = {};
  }

  roster() {
    // Return a deep copy with sorted arrays to prevent external mutation
    const copy = {};
    for (const grade in this._db) {
      copy[grade] = [...this._db[grade]].sort();
    }
    return copy;
  }

  add(name, grade) {
    // Remove student from any existing grade
    for (const g in this._db) {
      this._db[g] = this._db[g].filter(student => student !== name);
      // Clean up empty grades
      if (this._db[g].length === 0) {
        delete this._db[g];
      }
    }

    // Add student to the new grade
    if (!this._db[grade]) {
      this._db[grade] = [];
    }
    this._db[grade].push(name);
    this._db[grade].sort();
  }

  grade(gradeNumber) {
    if (!this._db[gradeNumber]) {
      return [];
    }
    // Return a sorted copy to prevent external mutation
    return [...this._db[gradeNumber]].sort();
  }
}

File: python/grade-school/grade_school.py
class School:
    def __init__(self):
        self._grades = {}  # grade_number -> set of names
        self._all_students = set()  # all unique student names
        self._add_results = []  # list of bools for add attempts

    def add_student(self, name, grade):
        if name in self._all_students:
            # Student already added somewhere
            self._add_results.append(False)
            return
        # New student
        self._all_students.add(name)
        if grade not in self._grades:
            self._grades[grade] = set()
        self._grades[grade].add(name)
        self._add_results.append(True)

    def roster(self):
        result = []
        for grade in sorted(self._grades):
            result.extend(sorted(self._grades[grade]))
        return result

    def grade(self, grade_number):
        if grade_number not in self._grades:
            return []
        return sorted(self._grades[grade_number])

    def added(self):
        return self._add_results
